cluster/juju/layers/kubernetes-master/reactive/kubernetes_master.py

```diff
@@ -1639,7 +1647,25 @@ def clear_cluster_tag_sent():
 
 
 @when_any('endpoint.aws.joined',
-          'endpoint.gcp.joined')
+          'endpoint.gcp.joined',
+          'endpoint.openstack.joined',
+          'endpoint.vsphere.joined',
+          'endpoint.azure.joined')
+@when_not('kubernetes-master.cloud.ready')
+def set_cloud_pending():
+    k8s_version = get_version('kube-apiserver')
+    k8s_1_11 = k8s_version >= (1, 11)
+    k8s_1_12 = k8s_version >= (1, 12)
+    vsphere_joined = is_state('endpoint.vsphere.joined')
+    azure_joined = is_state('endpoint.azure.joined')
+    if (vsphere_joined and not k8s_1_12) or (azure_joined and not k8s_1_11):
+        set_state('kubernetes-master.cloud.blocked')
+    set_state('kubernetes-master.cloud.pending')
```

Comment:
Could this be preempted by the `cloud_ready` handler? This handler might never get a chance to run, which would prevent `kubernetes-master.cloud.blocked` from being set.

Reply:
Ah, no, I missed that `cloud_ready` is gated on `endpoint.*.ready`, not `endpoint.*.joined`. This is probably OK.

Reply:
Actually, is it still a concern for interfaces that don't have a clear `joined` -> request integration -> `ready` flow? I'm thinking of openstack and vsphere in particular.